File: cubesat.c
// pps_host_libusb.c
#include <libusb-1.0/libusb.h>
#include <pthread.h>
#include <signal.h>
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <time.h>
#include <errno.h>

#define TX_FILE "/home/cubesat/Cubesat/Proj1/tx/waveforms/Hamming_burst_QPSK_msg_5Mhzfs_3sec.dat"
#define RX_DIR  "/home/cubesat/Cubesat/Proj1/rx/data/iq"

#define SAMPLE_RATE_HZ 5000000
#define CENTER_FREQ_HZ 10000000

#define AMP_ENABLE 0
#define ANTENNA_ENABLE 0
#define LNA_GAIN 16
#define VGA_GAIN 20
#define TXVGA_GAIN 20

#define CHUNK_SIZE 0x4000
#define READ_TIMEOUT_MS 20
#define WRITE_TIMEOUT_MS 20
#define STATUS_PRINT_SEC 2

#define VID 0x1D50
#define PID 0x6089

#define EP_IN  0x81
#define EP_OUT 0x02

#define VENDOR_OUT 0x40
#define VENDOR_IN  0xC0

#define HACKRF_VENDOR_REQUEST_SET_TRANSCEIVER_MODE 1
#define HACKRF_VENDOR_REQUEST_SAMPLE_RATE_SET      6
#define HACKRF_VENDOR_REQUEST_BASEBAND_FILTER_BW   7
#define HACKRF_VENDOR_REQUEST_SET_FREQ             16
#define HACKRF_VENDOR_REQUEST_AMP_ENABLE           17
#define HACKRF_VENDOR_REQUEST_SET_LNA_GAIN         19
#define HACKRF_VENDOR_REQUEST_SET_VGA_GAIN         20
#define HACKRF_VENDOR_REQUEST_SET_TXVGA_GAIN       21
#define HACKRF_VENDOR_REQUEST_ANTENNA_ENABLE       23

#define TRANSCEIVER_MODE_OFF 0

static volatile sig_atomic_t running = 1;

static pthread_mutex_t tx_lock = PTHREAD_MUTEX_INITIALIZER;
static pthread_mutex_t rx_lock = PTHREAD_MUTEX_INITIALIZER;

static uint64_t tx_bytes = 0;
static uint64_t rx_bytes = 0;

typedef struct {
    libusb_device_handle *dev;
    FILE *tx_fp;
} tx_ctx_t;

typedef struct {
    libusb_device_handle *dev;
    FILE *rx_fp;
} rx_ctx_t;

static void log_msg(const char *msg) {
    printf("%s\n", msg);
    fflush(stdout);
}

static int ctrl_out(libusb_device_handle *dev,
                    uint8_t request,
                    uint16_t value,
                    uint16_t index,
                    const unsigned char *data,
                    uint16_t length)
{
    return libusb_control_transfer(
        dev,
        VENDOR_OUT,
        request,
        value,
        index,
        (unsigned char *)data,
        length,
        1000
    );
}

static int ctrl_in(libusb_device_handle *dev,
                   uint8_t request,
                   uint16_t value,
                   uint16_t index,
                   unsigned char *data,
                   uint16_t length)
{
    return libusb_control_transfer(
        dev,
        VENDOR_IN,
        request,
        value,
        index,
        data,
        length,
        1000
    );
}

static int set_mode(libusb_device_handle *dev, uint8_t mode) {
    return ctrl_out(dev, HACKRF_VENDOR_REQUEST_SET_TRANSCEIVER_MODE, mode, 0, NULL, 0);
}

static int set_freq(libusb_device_handle *dev, uint32_t freq_hz) {
    uint32_t mhz = freq_hz / 1000000;
    uint32_t hz  = freq_hz % 1000000;
    unsigned char payload[8];

    memcpy(payload, &mhz, 4);
    memcpy(payload + 4, &hz, 4);

    return ctrl_out(dev, HACKRF_VENDOR_REQUEST_SET_FREQ, 0, 0, payload, sizeof(payload));
}

static int set_sample_rate(libusb_device_handle *dev, uint32_t freq_hz, uint32_t divider) {
    unsigned char payload[8];
    memcpy(payload, &freq_hz, 4);
    memcpy(payload + 4, &divider, 4);
    return ctrl_out(dev, HACKRF_VENDOR_REQUEST_SAMPLE_RATE_SET, 0, 0, payload, sizeof(payload));
}

static int set_baseband_filter_bw(libusb_device_handle *dev, uint32_t bw_hz) {
    uint16_t value = bw_hz & 0xFFFF;
    uint16_t index = (bw_hz >> 16) & 0xFFFF;
    return ctrl_out(dev, HACKRF_VENDOR_REQUEST_BASEBAND_FILTER_BW, value, index, NULL, 0);
}

static int set_amp_enable(libusb_device_handle *dev, bool enabled) {
    return ctrl_out(dev, HACKRF_VENDOR_REQUEST_AMP_ENABLE, enabled ? 1 : 0, 0, NULL, 0);
}

static int set_antenna_enable(libusb_device_handle *dev, bool enabled) {
    return ctrl_out(dev, HACKRF_VENDOR_REQUEST_ANTENNA_ENABLE, enabled ? 1 : 0, 0, NULL, 0);
}

static int set_lna_gain(libusb_device_handle *dev, uint16_t gain) {
    unsigned char resp[1];
    int ret = ctrl_in(dev, HACKRF_VENDOR_REQUEST_SET_LNA_GAIN, 0, gain, resp, 1);
    return ret < 0 ? ret : 0;
}

static int set_vga_gain(libusb_device_handle *dev, uint16_t gain) {
    unsigned char resp[1];
    int ret = ctrl_in(dev, HACKRF_VENDOR_REQUEST_SET_VGA_GAIN, 0, gain, resp, 1);
    return ret < 0 ? ret : 0;
}

static int set_txvga_gain(libusb_device_handle *dev, uint16_t gain) {
    unsigned char resp[1];
    int ret = ctrl_in(dev, HACKRF_VENDOR_REQUEST_SET_TXVGA_GAIN, 0, gain, resp, 1);
    return ret < 0 ? ret : 0;
}

static int configure_device(libusb_device_handle *dev) {
    int ret;

    ret = set_mode(dev, TRANSCEIVER_MODE_OFF);
    if (ret < 0) return ret;

    ret = set_sample_rate(dev, SAMPLE_RATE_HZ, 1);
    if (ret < 0) return ret;

    ret = set_baseband_filter_bw(dev, 5000000);
    if (ret < 0) return ret;

    ret = set_freq(dev, CENTER_FREQ_HZ);
    if (ret < 0) return ret;

    ret = set_amp_enable(dev, AMP_ENABLE);
    if (ret < 0) return ret;

    ret = set_antenna_enable(dev, ANTENNA_ENABLE);
    if (ret < 0) return ret;

    ret = set_lna_gain(dev, LNA_GAIN);
    if (ret < 0) return ret;

    ret = set_vga_gain(dev, VGA_GAIN);
    if (ret < 0) return ret;

    ret = set_txvga_gain(dev, TXVGA_GAIN);
    if (ret < 0) return ret;

    ret = set_mode(dev, TRANSCEIVER_MODE_OFF);
    if (ret < 0) return ret;

    return 0;
}

static void make_rx_path(char *out, size_t out_sz) {
    time_t now = time(NULL);
    struct tm tm_now;
    localtime_r(&now, &tm_now);

    mkdir(RX_DIR, 0777);

    snprintf(out, out_sz,
             RX_DIR "/pps_rx_%04d%02d%02d_%02d%02d%02d.dat",
             tm_now.tm_year + 1900,
             tm_now.tm_mon + 1,
             tm_now.tm_mday,
             tm_now.tm_hour,
             tm_now.tm_min,
             tm_now.tm_sec);
}

static void *tx_worker(void *arg) {
    tx_ctx_t *ctx = (tx_ctx_t *)arg;
    unsigned char buf[CHUNK_SIZE];

    while (running) {
        size_t got = fread(buf, 1, CHUNK_SIZE, ctx->tx_fp);

        if (got == 0) {
            fseek(ctx->tx_fp, 0, SEEK_SET);
            got = fread(buf, 1, CHUNK_SIZE, ctx->tx_fp);
            if (got == 0) {
                usleep(10000);
                continue;
            }
        }

        if (got < CHUNK_SIZE) {
            fseek(ctx->tx_fp, 0, SEEK_SET);
            size_t remain = CHUNK_SIZE - got;
            size_t got2 = fread(buf + got, 1, remain, ctx->tx_fp);
            got += got2;

            if (got < CHUNK_SIZE) {
                memset(buf + got, 0, CHUNK_SIZE - got);
                got = CHUNK_SIZE;
            }
        }

        int transferred = 0;
        int ret = libusb_bulk_transfer(
            ctx->dev,
            EP_OUT,
            buf,
            CHUNK_SIZE,
            &transferred,
            WRITE_TIMEOUT_MS
        );

        if (ret == 0) {
            pthread_mutex_lock(&tx_lock);
            tx_bytes += (uint64_t)transferred;
            pthread_mutex_unlock(&tx_lock);
        } else if (ret == LIBUSB_ERROR_TIMEOUT) {
            // Normal when firmware is not in TX mode.
            continue;
        } else {
            fprintf(stderr, "[TX] USB error: %s\n", libusb_error_name(ret));
            running = 0;
            return NULL;
        }
    }

    return NULL;
}

static void *rx_worker(void *arg) {
    rx_ctx_t *ctx = (rx_ctx_t *)arg;
    unsigned char buf[CHUNK_SIZE];

    while (running) {
        int transferred = 0;
        int ret = libusb_bulk_transfer(
            ctx->dev,
            EP_IN,
            buf,
            CHUNK_SIZE,
            &transferred,
            READ_TIMEOUT_MS
        );

        if (ret == 0) {
            if (transferred > 0) {
                fwrite(buf, 1, transferred, ctx->rx_fp);
                pthread_mutex_lock(&rx_lock);
                rx_bytes += (uint64_t)transferred;
                pthread_mutex_unlock(&rx_lock);
            }
        } else if (ret == LIBUSB_ERROR_TIMEOUT) {
            // Normal when firmware is not in RX mode.
            continue;
        } else {
            fprintf(stderr, "[RX] USB error: %s\n", libusb_error_name(ret));
            running = 0;
            return NULL;
        }
    }

    return NULL;
}

static void *status_worker(void *arg) {
    (void)arg;

    uint64_t last_tx = 0;
    uint64_t last_rx = 0;

    while (running) {
        sleep(STATUS_PRINT_SEC);

        pthread_mutex_lock(&tx_lock);
        uint64_t cur_tx = tx_bytes;
        pthread_mutex_unlock(&tx_lock);

        pthread_mutex_lock(&rx_lock);
        uint64_t cur_rx = rx_bytes;
        pthread_mutex_unlock(&rx_lock);

        uint64_t dtx = cur_tx - last_tx;
        uint64_t drx = cur_rx - last_rx;
        last_tx = cur_tx;
        last_rx = cur_rx;

        printf("[STATUS] total_tx=%llu B total_rx=%llu B | tx_rate=%.1f B/s | rx_rate=%.1f B/s\n",
               (unsigned long long)cur_tx,
               (unsigned long long)cur_rx,
               (double)dtx / STATUS_PRINT_SEC,
               (double)drx / STATUS_PRINT_SEC);
        fflush(stdout);
    }

    return NULL;
}

static void handle_signal(int sig) {
    running = 0;
    fprintf(stderr, "\nStopping on signal %d...\n", sig);
}

int main(void) {
    int ret;
    libusb_context *usb_ctx = NULL;
    libusb_device_handle *dev = NULL;

    if (access(TX_FILE, F_OK) != 0) {
        fprintf(stderr, "TX file not found: %s\n", TX_FILE);
        return 1;
    }

    char rx_path[512];
    make_rx_path(rx_path, sizeof(rx_path));

    printf("TX file:      %s\n", TX_FILE);
    printf("RX output:    %s\n", rx_path);
    printf("Sample rate:  %d Hz\n", SAMPLE_RATE_HZ);
    printf("Center freq:  %d Hz\n", CENTER_FREQ_HZ);

    ret = libusb_init(&usb_ctx);
    if (ret < 0) {
        fprintf(stderr, "libusb_init failed: %s\n", libusb_error_name(ret));
        return 1;
    }

    dev = libusb_open_device_with_vid_pid(usb_ctx, VID, PID);
    if (!dev) {
        fprintf(stderr, "HackRF not found\n");
        libusb_exit(usb_ctx);
        return 1;
    }

    if (libusb_kernel_driver_active(dev, 0) == 1) {
        libusb_detach_kernel_driver(dev, 0);
    }

    ret = libusb_claim_interface(dev, 0);
    if (ret < 0) {
        fprintf(stderr, "claim_interface failed: %s\n", libusb_error_name(ret));
        libusb_close(dev);
        libusb_exit(usb_ctx);
        return 1;
    }

    ret = configure_device(dev);
    if (ret < 0) {
        fprintf(stderr, "configure_device failed: %s\n", libusb_error_name(ret));
        libusb_release_interface(dev, 0);
        libusb_close(dev);
        libusb_exit(usb_ctx);
        return 1;
    }

    log_msg("HackRF configured. Firmware remains in OFF mode until PPS scheduler switches modes.");

    signal(SIGINT, handle_signal);
    signal(SIGTERM, handle_signal);

    FILE *tx_fp = fopen(TX_FILE, "rb");
    if (!tx_fp) {
        perror("fopen TX_FILE");
        libusb_release_interface(dev, 0);
        libusb_close(dev);
        libusb_exit(usb_ctx);
        return 1;
    }

    FILE *rx_fp = fopen(rx_path, "wb");
    if (!rx_fp) {
        perror("fopen RX output");
        fclose(tx_fp);
        libusb_release_interface(dev, 0);
        libusb_close(dev);
        libusb_exit(usb_ctx);
        return 1;
    }

    tx_ctx_t tx_ctx = { .dev = dev, .tx_fp = tx_fp };
    rx_ctx_t rx_ctx = { .dev = dev, .rx_fp = rx_fp };

    pthread_t tx_thread, rx_thread, st_thread;

    pthread_create(&tx_thread, NULL, tx_worker, &tx_ctx);
    pthread_create(&rx_thread, NULL, rx_worker, &rx_ctx);
    pthread_create(&st_thread, NULL, status_worker, NULL);

    while (running) {
        usleep(200000);
        fflush(rx_fp);
    }

    pthread_join(tx_thread, NULL);
    pthread_join(rx_thread, NULL);
    pthread_join(st_thread, NULL);

    set_mode(dev, TRANSCEIVER_MODE_OFF);

    fclose(tx_fp);
    fclose(rx_fp);

    libusb_release_interface(dev, 0);
    libusb_close(dev);
    libusb_exit(usb_ctx);

    log_msg("Done.");
    return 0;
}

File: cubesat.py
#!/usr/bin/env python3
import os
import sys
import time
import signal
import struct
import threading
from datetime import datetime

import usb.core
import usb.util

# ============================================================
# User config
# ============================================================
TX_FILE = "/home/cubesat/CubeSat/Tx/waveforms/Hamming_burst_QPSK_msg_5Mhzfs_3sec.dat"
RX_DIR  = "/home/cubesat/CubeSat/Tx/data/iq"

SAMPLE_RATE_HZ = 5_000_000
CENTER_FREQ_HZ = 10_000_000

AMP_ENABLE = 0
ANTENNA_ENABLE = 0
LNA_GAIN = 16      # RX IF gain
VGA_GAIN = 20      # RX BB gain
TXVGA_GAIN = 20    # TX IF gain

CHUNK_SIZE = 0x4000         # 16384 bytes, matches your firmware USB_TRANSFER_SIZE
READ_TIMEOUT_MS = 20
WRITE_TIMEOUT_MS = 20
STATUS_PRINT_SEC = 2.0

# HackRF USB IDs
VID = 0x1D50
PID = 0x6089

# Bulk endpoints
EP_IN = 0x81   # device -> host
EP_OUT = 0x02  # host -> device

# ============================================================
# HackRF vendor requests
# ============================================================
VENDOR_OUT = 0x40
VENDOR_IN  = 0xC0

HACKRF_VENDOR_REQUEST_SET_TRANSCEIVER_MODE = 1
HACKRF_VENDOR_REQUEST_SAMPLE_RATE_SET      = 6
HACKRF_VENDOR_REQUEST_BASEBAND_FILTER_BW   = 7
HACKRF_VENDOR_REQUEST_SET_FREQ             = 16
HACKRF_VENDOR_REQUEST_AMP_ENABLE           = 17
HACKRF_VENDOR_REQUEST_SET_LNA_GAIN         = 19
HACKRF_VENDOR_REQUEST_SET_VGA_GAIN         = 20
HACKRF_VENDOR_REQUEST_SET_TXVGA_GAIN       = 21
HACKRF_VENDOR_REQUEST_ANTENNA_ENABLE       = 23

TRANSCEIVER_MODE_OFF = 0

running = True
tx_bytes = 0
rx_bytes = 0
tx_lock = threading.Lock()
rx_lock = threading.Lock()


def log(msg: str) -> None:
    print(msg, flush=True)


def is_timeout(exc: Exception) -> bool:
    s = str(exc).lower()
    return "timed out" in s or "timeout" in s


def ctrl_out(dev, request: int, value: int = 0, index: int = 0, data: bytes = b"") -> None:
    dev.ctrl_transfer(VENDOR_OUT, request, value, index, data, timeout=1000)


def ctrl_in(dev, request: int, value: int = 0, index: int = 0, length: int = 1) -> bytes:
    data = dev.ctrl_transfer(VENDOR_IN, request, value, index, length, timeout=1000)
    return bytes(data)


def set_mode(dev, mode: int) -> None:
    ctrl_out(dev, HACKRF_VENDOR_REQUEST_SET_TRANSCEIVER_MODE, mode, 0, b"")


def set_freq(dev, freq_hz: int) -> None:
    mhz = freq_hz // 1_000_000
    hz  = freq_hz % 1_000_000
    payload = struct.pack("<II", mhz, hz)
    ctrl_out(dev, HACKRF_VENDOR_REQUEST_SET_FREQ, 0, 0, payload)


def set_sample_rate(dev, freq_hz: int, divider: int = 1) -> None:
    payload = struct.pack("<II", freq_hz, divider)
    ctrl_out(dev, HACKRF_VENDOR_REQUEST_SAMPLE_RATE_SET, 0, 0, payload)


def set_baseband_filter_bw(dev, bw_hz: int) -> None:
    value = bw_hz & 0xFFFF
    index = (bw_hz >> 16) & 0xFFFF
    ctrl_out(dev, HACKRF_VENDOR_REQUEST_BASEBAND_FILTER_BW, value, index, b"")


def set_amp_enable(dev, enabled: bool) -> None:
    ctrl_out(dev, HACKRF_VENDOR_REQUEST_AMP_ENABLE, 1 if enabled else 0, 0, b"")


def set_antenna_enable(dev, enabled: bool) -> None:
    ctrl_out(dev, HACKRF_VENDOR_REQUEST_ANTENNA_ENABLE, 1 if enabled else 0, 0, b"")


def set_lna_gain(dev, gain: int) -> None:
    resp = ctrl_in(dev, HACKRF_VENDOR_REQUEST_SET_LNA_GAIN, 0, gain, 1)
    if not resp:
        raise RuntimeError("Failed to set LNA gain")


def set_vga_gain(dev, gain: int) -> None:
    resp = ctrl_in(dev, HACKRF_VENDOR_REQUEST_SET_VGA_GAIN, 0, gain, 1)
    if not resp:
        raise RuntimeError("Failed to set VGA gain")


def set_txvga_gain(dev, gain: int) -> None:
    resp = ctrl_in(dev, HACKRF_VENDOR_REQUEST_SET_TXVGA_GAIN, 0, gain, 1)
    if not resp:
        raise RuntimeError("Failed to set TXVGA gain")


def configure_device(dev) -> None:
    # Host config only. Firmware owns PPS timing + RX/TX switching.
    set_mode(dev, TRANSCEIVER_MODE_OFF)
    set_sample_rate(dev, SAMPLE_RATE_HZ, 1)
    set_baseband_filter_bw(dev, 5_000_000)
    set_freq(dev, CENTER_FREQ_HZ)
    set_amp_enable(dev, bool(AMP_ENABLE))
    set_antenna_enable(dev, bool(ANTENNA_ENABLE))
    set_lna_gain(dev, LNA_GAIN)
    set_vga_gain(dev, VGA_GAIN)
    set_txvga_gain(dev, TXVGA_GAIN)
    set_mode(dev, TRANSCEIVER_MODE_OFF)


def make_rx_path() -> str:
    os.makedirs(RX_DIR, exist_ok=True)
    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    return os.path.join(RX_DIR, f"pps_rx_{stamp}.dat")


def tx_worker(dev, tx_fp):
    global running, tx_bytes

    while running:
        chunk = tx_fp.read(CHUNK_SIZE)

        if not chunk:
            tx_fp.seek(0)
            chunk = tx_fp.read(CHUNK_SIZE)
            if not chunk:
                time.sleep(0.01)
                continue

        if len(chunk) < CHUNK_SIZE:
            tx_fp.seek(0)
            remaining = CHUNK_SIZE - len(chunk)
            chunk += tx_fp.read(remaining)
            if len(chunk) < CHUNK_SIZE:
                chunk += b"\x00" * (CHUNK_SIZE - len(chunk))

        try:
            written = dev.write(EP_OUT, chunk, timeout=WRITE_TIMEOUT_MS)
            with tx_lock:
                tx_bytes += int(written)
        except usb.core.USBError as e:
            if is_timeout(e):
                # Normal when firmware is not currently in TX mode
                continue
            log(f"[TX] USB error: {e}")
            running = False
            return
        except Exception as e:
            log(f"[TX] fatal error: {e}")
            running = False
            return


def rx_worker(dev, rx_fp):
    global running, rx_bytes

    while running:
        try:
            data = dev.read(EP_IN, CHUNK_SIZE, timeout=READ_TIMEOUT_MS)
            raw = bytes(data)
            if raw:
                rx_fp.write(raw)
                with rx_lock:
                    rx_bytes += len(raw)
        except usb.core.USBError as e:
            if is_timeout(e):
                # Normal when firmware is not currently in RX mode
                continue
            log(f"[RX] USB error: {e}")
            running = False
            return
        except Exception as e:
            log(f"[RX] fatal error: {e}")
            running = False
            return


def status_worker():
    global running
    last_tx = 0
    last_rx = 0

    while running:
        time.sleep(STATUS_PRINT_SEC)

        with tx_lock:
            cur_tx = tx_bytes
        with rx_lock:
            cur_rx = rx_bytes

        dtx = cur_tx - last_tx
        drx = cur_rx - last_rx
        last_tx = cur_tx
        last_rx = cur_rx

        log(
            f"[STATUS] total_tx={cur_tx} B total_rx={cur_rx} B "
            f"| tx_rate={dtx / STATUS_PRINT_SEC:.1f} B/s "
            f"| rx_rate={drx / STATUS_PRINT_SEC:.1f} B/s"
        )


def handle_signal(signum, frame):
    global running
    running = False
    log(f"\nStopping on signal {signum}...")


def main():
    global running

    if not os.path.isfile(TX_FILE):
        raise FileNotFoundError(f"TX file not found: {TX_FILE}")

    rx_path = make_rx_path()

    log(f"TX file:      {TX_FILE}")
    log(f"RX output:    {rx_path}")
    log(f"Sample rate:  {SAMPLE_RATE_HZ} Hz")
    log(f"Center freq:  {CENTER_FREQ_HZ} Hz")

    dev = usb.core.find(idVendor=VID, idProduct=PID)
    if dev is None:
        raise RuntimeError("HackRF not found")

    try:
        if dev.is_kernel_driver_active(0):
            dev.detach_kernel_driver(0)
    except (NotImplementedError, usb.core.USBError):
        pass

    dev.set_configuration()
    usb.util.claim_interface(dev, 0)

    configure_device(dev)
    log("HackRF configured. Firmware remains in OFF mode until PPS scheduler switches modes.")

    signal.signal(signal.SIGINT, handle_signal)
    signal.signal(signal.SIGTERM, handle_signal)

    with open(TX_FILE, "rb") as tx_fp, open(rx_path, "wb") as rx_fp:
        tx_thread = threading.Thread(target=tx_worker, args=(dev, tx_fp), daemon=True)
        rx_thread = threading.Thread(target=rx_worker, args=(dev, rx_fp), daemon=True)
        st_thread = threading.Thread(target=status_worker, daemon=True)

        tx_thread.start()
        rx_thread.start()
        st_thread.start()

        try:
            while running:
                time.sleep(0.2)
                rx_fp.flush()
        finally:
            running = False
            time.sleep(0.2)
            try:
                set_mode(dev, TRANSCEIVER_MODE_OFF)
            except Exception:
                pass
            try:
                usb.util.release_interface(dev, 0)
            except Exception:
                pass
            usb.util.dispose_resources(dev)

    log("Done.")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        log(f"Fatal: {e}")
        sys.exit(1)